### PR TITLE
Extending ObservableQuery to refactor implementations.

### PR DESCRIPTION
**Pull request by tmeasday**
#554 TODOs:

- [x] Add `.setVariables()`
- [x] Add a syncronous version of `.current()`
- [ ] Refactor `react-apollo` to use this (I don't think we should merge this until that is done).

Standard PR TODOs:

- [x] If this PR is a new feature, reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests
- [ ] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [ ] Update CHANGELOG.md with your change
- [ ] Add your name and email to the AUTHORS file (optional)

Note, this is still a WIP, but I'd like some comments, I'll write some inline.

### CODE REVIEW

File: src/ObservableQuery.ts [hunk 1]
```diff
@@ -91,25 +108,42 @@ export class ObservableQuery extends Observable<ApolloQueryResult> {
     };
     super(subscriberFunction);
     this.options = options;
+    this.variables = this.options.variables || {};
```
Comment (tmeasday) [inline src/ObservableQuery.ts]: This means variables is never `null` any more. This seems logically fine to me ("empty set" == no variables), and seems to simplify a bunch of other code.

File: src/ObservableQuery.ts [hunk 2]
```diff
@@ -121,8 +155,8 @@ export class ObservableQuery extends Observable<ApolloQueryResult> {
             combinedOptions = fetchMoreOptions;
           } else {
             // fetch the same query with a possibly new variables
-            const variables = this.options.variables || fetchMoreOptions.variables ?
-              assign({}, this.options.variables, fetchMoreOptions.variables) : undefined;
+            const variables = this.variables || fetchMoreOptions.variables ?
```
Comment (tmeasday) [inline src/ObservableQuery.ts]: This logic seems wrong to me. Why do we ignore `fetchMoreOptions.variables` if `this.variables` is defined?

Any idea @helfer?

File: src/ObservableQuery.ts (hunk 1 above)
Comment (stubailo) [inline src/ObservableQuery.ts]: I'm OK with requiring that variables are never `null` internally.

File: src/ObservableQuery.ts (hunk 2 above)
Comment (tmeasday) [inline src/ObservableQuery.ts]: Oh, I just realised that `||` has higher precedence than `? :`. Uggh. I'm going to simplify given the "variables is never null" logic above.

File: test/mocks/mockNetworkInterface.ts [hunk 3]
```diff
@@ -199,7 +199,7 @@ function requestToKey(request: ParsedRequest): string {
   const queryString = request.query && print(request.query);
 
   return JSON.stringify({
-    variables: request.variables,
+    variables: request.variables || {},
```
Comment (tmeasday) [inline test/mocks/mockNetworkInterface.ts]: See note about variables not being null above.

File: src/ObservableQuery.ts (hunk 1 above)
Comment (tmeasday) [inline src/ObservableQuery.ts]: I'm not sure it's enforced at this point, but maybe we should do that.

File: src/ObservableQuery.ts [hunk 4]
```diff
@@ -204,4 +237,17 @@ export class ObservableQuery extends Observable<ApolloQueryResult> {
       });
     });
   }
+
+  public currentResult(): ApolloQueryResult {
+    try {
+      const { previousResult } = this.queryManager.getQueryWithPreviousResult(this.queryId);
+      return { data: previousResult, loading: false };
+    } catch (e) {
+      if (e && e.extraInfo.isFieldError) {
```
Comment (tmeasday) [inline src/ObservableQuery.ts]: I don't think we should use `try { } catch` for this, but it's not clear what the best way to determine if a query is in flight from the query manager is.

File: test/ObservableQuery.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  describe('currentResult', () => {
+    it('returns the current query status immediately', (done) => {
+      const observable: ObservableQuery = mockWatchQuery({
+        request: { query, variables },
+        result: { data: dataOne },
+        delay: 100,
+      });
+
+      // XXX: should I need to subscribe for this to work?
+      observable.subscribe({ next() {} }); // tslint:disable-line
```
Comment (tmeasday) [inline test/ObservableQuery.ts]: I don't think you should need to subscribe to a query to be able to run `observable.currentResult()`.

**Comment by tmeasday** [comment]
@helfer can you take a look at this? I'd like to figure out the best way to implement the `currentResult()` functionality, the current technique seems bad to me.

File: src/ObservableQuery.ts (hunk 4 above)
Comment (stubailo) [inline src/ObservableQuery.ts]: Use the `queryId` to look it up in the store, under `queries`.

Comment (tmeasday) [inline src/ObservableQuery.ts]: So I should directly read from the store in `ObservableQuery`? Or should I add something to `QueryManager` to do this?

Comment (stubailo) [inline src/ObservableQuery.ts]: Hard to say. At some point, I thought maybe QueryManager would be the abstraction layer between the store and everything else, but i'm not sure if that's the case anymore. I'd say do what makes sense and we can refactor later.

**Comment by stubailo** [comment]
We've been thinking of ditching the current observable API: https://github.com/apollostack/apollo-client/issues/622

We've been running into a lot of problems with this API with subscriptions as well, it's both too complicated and not flexible enough for our needs IMO.

**Comment by stubailo** [comment]
BTW, any way to get `setVariables` and the other thing into separate PRs?

**Comment by tmeasday** [comment]
@stubailo ok, I'll split off the two PRs

**Comment by tmeasday** [comment]
Ok, https://github.com/apollostack/apollo-client/pull/634 / https://github.com/apollostack/apollo-client/pull/635.

Not sure what to do with this PR?

**Pull request closed without merging**